### PR TITLE
gnrc_sock: provide port for sock_ip and sock_udp

### PR DESCRIPTION
**Pull request by miri64**
This PR provides the port for `sock_ip` and `sock_udp` (see #5758) to GNRC + tests.

Depends on ~~#5526~~ (`mbox` support for GNRC), ~~#5749~~ (fix to UDP so the tests are able to run), and ~~#5758~~ (the `sock` header files).

### CODE REVIEW

**Comment by miri64** [comment]
@brummer-simon this PR might also be of interest for you to give you an idea for how to port `sock_tcp` using `mbox` ;-).

**Comment by miri64** [comment]
Rebased due to changes in #5526.

**Comment by kaspar030** [comment]
sock is merged, pls rebase

**Comment by miri64** [comment]
Can we merge the dependencies first please? Also, while working on lwIP I realized I need to adapt the tests a little ;-).

**Comment by miri64** [comment]
Rebased to current dependencies and master.

**Comment by miri64** [comment]
Applied the changes I spoke about too.

**Comment by miri64** [comment]
Rebased to current master, since #5884 was merged.

**Comment by miri64** [comment]
While working on #5802 I realized that their raw sock equivalent require the IP header to be provided. This seems to be the behavior with POSIX sockets, too (see e.g. [this tutorial](http://www.tenouk.com/Module43a.html)). I did not implement it like that in `gnrc_sock_ip`, because the API was conceptualized under a different premise (the header is build by the API). However that would make the implementation of the socket wrapper much more complicated. So how do we handle this?

**Comment by miri64** [comment]
Nvm... lwIP isn't consistent in this itself... and so is the information about raw sockets themselves..

**Comment by miri64** [comment]
Needs some small adjustments due to #5929 being merged.

**Comment by miri64** [comment]
Rebased and squashed to current master to work on this ASAP.

**Comment by miri64** [comment]
Adapted for #5929.

**Comment by miri64** [comment]
Rebased to current #5526.

**Comment by haukepetersen** [comment]
From my side everything looks in order. Tests succeed and I successfully tested the `sock_udp` implementation under native. ACK once squashed and Murdock is happy.

**Comment by miri64** [comment]
Squashed

**Comment by miri64** [comment]
Rebased to current master.

**Comment by miri64** [comment]
Ooops, didn't adapt it to the pseudo-module change in #5526. Done now.

**Comment by cgundogan** [comment]
failed to download the `tiny-asn1` pkg, retriggering murdock

**Comment by cgundogan** [comment]
Murdock still fails @miri64

**Comment by miri64** [comment]
Fixed, rebased and squashed.

**Comment by miri64** [comment]
Okay, `cppcheck` is really acting weird on this lines... first it complains about the "redundant" check (dereferencing `sock` after I checked if it is `NULL` in an `||` connection, which thanks to lazy evaluation should be fine), when I suppress it it complains that there is a `NULL` pointer dereference at that line (no it isn't because lazy evaluation) and when I write it out for it (`(sock == NULL) || ((sock != NULL) && (sock->foobar == ...))`) it is complaining that the additional check for `(sock != NULL)` isn't necessary, since `(sock == NULL) || (sock->foobar == ...)` is equivalent (which is what was there in the first place...) so I just suppressed everything on this line. Since my tests are covering this line and they are not crashing I guess we are fine and `cppcheck` just is weird. @cgundogan could you re-ACK please? The changes are in https://github.com/RIOT-OS/RIOT/pull/5772/files#diff-71ddba2942574ad9a8d73c8de31b2230R188 and https://github.com/RIOT-OS/RIOT/pull/5772/files#diff-97c7bd9404c4bc328009313d1665dfe5R150 respectively.

**Comment by miri64** [comment]
Murdock is finally happy!

**Comment by haukepetersen** [comment]
I am good with the changes (so re-ACK) and Murdock is happy -> go

@miri64 Congrats, this is big step towards usability of GNRC!

File: sys/net/gnrc/sock/ip/gnrc_sock_ip.c [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+        memcpy(&rem, &sock->remote, sizeof(rem));
+    }
+    else {
+        memcpy(&rem, remote, sizeof(rem));
+    }
+    if ((remote != NULL) && (remote->family == AF_UNSPEC) &&
+        (sock->remote.family != AF_UNSPEC)) {
+        /* remote was set on create so take its family */
+        rem.family = sock->remote.family;
+    }
```
Comment (keestux) [inline sys/net/gnrc/sock/ip/gnrc_sock_ip.c:175]: @miri64 
The assert above indicates that `remote` or `sock` must be non-NULL.
At this if statement we check that `remote` is not NULL. But in that case it is possible that `sock` is NULL. We don't check that before using `sock`.

Comment (miri64) [reply sys/net/gnrc/sock/ip/gnrc_sock_ip.c:175]: Confirmed and fixed in https://github.com/RIOT-OS/RIOT/pull/14963

File: sys/net/gnrc/sock/udp/gnrc_sock_udp.c [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+    if ((sock->remote.family != AF_UNSPEC) &&  /* check remote end-point if set */
+        ((sock->remote.port != byteorder_ntohs(hdr->src_port)) ||
+        /* We only have IPv6 for now, so just comparing the whole end point
+         * should suffice */
+        ((memcmp(&sock->remote.addr, &ipv6_addr_unspecified,
+                 sizeof(ipv6_addr_t)) != 0) &&
+         (memcmp(&sock->remote.addr, &tmp.addr, sizeof(ipv6_addr_t)) != 0)))) {
+        gnrc_pktbuf_release(pkt);
+        return -EPROTO;
+    }
```
Comment (benpicco) [inline sys/net/gnrc/sock/udp/gnrc_sock_udp.c:156]: Why is this condition needed?
This enforces that if the socket was previously used for sending, received packets must come from the same address.

But there is nothing that says a socket must be bound to the *remote* address, but there is even a test for this.

Why is this there? This makes it impossible to use a socket to send to a multicast address and receive unicast responses.

**Pull request merged**